scikeras/utils/transformers.py

```diff
@@ -0,0 +1,189 @@
+                    "Expected a 1D array for `X`." f" Got a {X.ndim}D array instead."
+                )
+            X = X.reshape(-1, 1)
+        return X
+
+    def inverse_transform(self, X: np.ndarray) -> np.ndarray:
+        if self.should_transform_:
+            if X.ndim != 2:
+                raise ValueError(
+                    f"Expected a 2D array for `X`, but got a"
+                    f" {X.ndim}D array instead."
+                )
+            if X.shape[1] != 1:
+                raise ValueError(
+                    f"Expected `X.shape[1] == 1`"
+                    f" but got `X.shape[1] == {X.shape}` instead."
+                )
+            X = np.squeeze(X, axis=1)
+        return X
+
+
+class ClassifierLabelEncoder(BaseEstimator, TransformerMixin):
+    """Default target transformer for KerasClassifier.
+    """
+
+    def __init__(self, loss=None):
+        self.loss = loss
+
+    def fit(self, X: np.ndarray) -> "ClassifierLabelEncoder":
+        y = X  # rename for clarity, the input is always expected to be a target `y`
+        target_type = type_of_target(y)
+        encoders = {
+            "binary": make_pipeline(
+                Ensure2DTransformer(), OrdinalEncoder(dtype=np.float32),
+            ),
+            "multiclass": make_pipeline(
+                Ensure2DTransformer(), OrdinalEncoder(dtype=np.float32),
+            ),
+            "multiclass-multioutput": FunctionTransformer(),
+            "multilabel-indicator": FunctionTransformer(),
+        }
+        if is_categorical_crossentropy(self.loss):
+            encoders["multiclass"] = make_pipeline(
+                Ensure2DTransformer(), OneHotEncoder(sparse=False, dtype=np.float32),
+            )
+        if target_type not in encoders:
+            raise ValueError(
+                f"Unknown label type: {target_type}."
+                "\n\nTo implement support, subclass KerasClassifier and override"
+                " `target_transformer` with a transformer that supports this"
+                " label type."
+            )
+        self.final_encoder_ = encoders[target_type].fit(y)
+
+        if (
+            target_type == "multilabel-indicator"
+            and y.min() == 0
+            and (y.sum(axis=1) == 1).all()
+        ):
+            target_type = "multiclass-onehot"
+
+        self.n_outputs_ = 1
+        self.model_n_outputs_ = 1
+        self.y_dtype_ = y.dtype
+        self._target_type = target_type
+
+        if target_type in ("binary", "multiclass"):
+            self.classes_ = self.final_encoder_[1].categories_[0]
+            self.n_classes_ = self.classes_.size
+        elif target_type in ("multiclass-onehot", "multilabel-indicator"):
+            self.classes_ = np.arange(0, y.shape[1])
+            self.n_classes_ = y.shape[1]
+        elif target_type == "multiclass-multioutput":
+            self.classes_ = [np.unique(y[:, j]) for j in range(y.shape[1])]
+            self.n_classes_ = len(self.n_classes_)
```

Comment:
I think this is valid for single-output models. I don't see any way you could create a single output in Keras that does not respect this. The only counter examples I can think of are strictly multi-output, eg:

```python3
array([[0, 1, 0],
       [1, 0, 1],
       [1, 0, 2]])
```
This would be classified `"multiclass-multioutput"`. The first two columns could be one-hot encoded targets for one input while the third is a categorical target for an output with `sparse_categorical_crossentropy` loss.

Reply:
Is SciKeras supporting `multiclass-multioutput` in KerasClassifier? Or is SciKeras not supporting multiple output classifiers with multiple losses?

Reply:
And where is `n_classes_` defined?

Reply:
> Is SciKeras supporting `multiclass-multioutput` in KerasClassifier? Or is SciKeras not supporting multiple output classifiers with multiple losses?

We decided that even though we do not support it (as in, do not do any data processing for it) but we still "allow" it by passing through the data without any transformation or errors. So the question is then what do we set these attributes to, or do we leave them out?

Reply:
> And where is `n_classes_` defined?

I'm a bit confused by this question. Are you asking for the formal definition of what it should mean, or are you asking where in the code it's defined?

If it's the former, the "definition" is to use it how sklearn estimators use it, for example in [DecisionTreeClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html#sklearn.tree.DecisionTreeClassifier).

Reply:
> And where is n_classes_ defined?

I mean in the line `self.n_classes_ = len(self.n_classes_)`. Do you mean `self.n_classes_ = len(self.classes_)`? I'm not seeing the definition of `self.n_classes_` used in `len(self.n_classes_)`.

> So the question is then what do we set these attributes to, or do we leave them out?

I think that's up to the user. I'd leave them out in this implementation. They can always subclass `ClassifierLabelEncoder` and set those attributes themselves:

``` python
class Custom(ClassifierLabelEncoder):
    def fit(self, X, y, **kwargs):
        ret = super().fit(X, y, **kwargs)
        self.n_classes_ = foo(X, y)
        self.classes_ = bar(X, y)
```

Reply:
I agree, but does that mean that `self.n_classes_ = None` or that `self.n_classes_` never gets set for that type?

I lean towards the former, I don't like changing the interface of the transformers (i.e. the keys in the `get_metadata` dict) based on the internal implementation.

Reply:
> I'm not seeing the definition of `self.n_classes_` used in `len(self.n_classes_)`

🤦 you're right, I see what you mean.

Reply:
> I lean towards the former, I don't like changing the interface of the transformers (i.e. the keys in the get_metadata dict) based on the internal implementation.

:+1:

Reply:
Implemented

```python3
elif target_type == "multiclass-multioutput":
    self.classes_ = None
    self.n_classes_ = None
```